From cc266a6c28f7b78473c15ec46047bb7a9bcf414a Mon Sep 17 00:00:00 2001
From: Tim S <tim.stahlhut@gmail.com>
Date: Tue, 9 Apr 2019 08:16:18 -0400
Subject: Fix gcc/ada/raise-gcc.c

Combined parts of two later patch files SVN 247301 and 247309
into single patch.

---
 gcc/ada/gcc-interface/Make-lang.in | 13 ++++++-----
 gcc/ada/raise-gcc.c                | 37 +++++++++++++++++++++++++-----
 2 files changed, 38 insertions(+), 12 deletions(-)

diff --git a/gcc/ada/gcc-interface/Make-lang.in b/gcc/ada/gcc-interface/Make-lang.in
index 10c865f457d..fa04e80a52d 100644
--- a/gcc/ada/gcc-interface/Make-lang.in
+++ b/gcc/ada/gcc-interface/Make-lang.in
@@ -99,8 +99,6 @@ ADA_TOOLS=gnatbind gnatchop gnat gnatkr gnatlink gnatls gnatmake \
 ada-warn = $(ADA_CFLAGS) $(filter-out -pedantic, $(STRICT_WARN))
 # Unresolved warnings in specific files.
 ada/adaint.o-warn = -Wno-error
-# For unwind-pe.h
-CFLAGS-ada/raise-gcc.o += -I$(srcdir)/../libgcc -Iinclude
 
 ada/%.o: ada/gcc-interface/%.c
 	$(COMPILE) $<
@@ -611,17 +609,20 @@ TARGET_ADA_SRCS =
 
 # Select the right s-excmac according to exception layout (Itanium or arm)
 host_cpu=$(word 1, $(subst -, ,$(host)))
-EH_MECHANISM=-gcc
+EH_MECHANISM=gcc
 ifeq ($(strip $(filter-out arm%,$(host_cpu))),)
-EH_MECHANISM=-arm
+EH_MECHANISM=arm
 endif
 
+# For unwind-pe.h
+CFLAGS-ada/raise-gcc.o += -I$(srcdir)/../libgcc -DEH_MECHANISM_$(EH_MECHANISM)
+
 ada/s-excmac.o: ada/s-excmac.ads ada/s-excmac.adb
 
-ada/s-excmac.ads: $(srcdir)/ada/s-excmac$(EH_MECHANISM).ads
+ada/s-excmac.ads: $(srcdir)/ada/s-excmac-$(EH_MECHANISM).ads
 	$(CP) $< $@
 
-ada/s-excmac.adb: $(srcdir)/ada/s-excmac$(EH_MECHANISM).adb
+ada/s-excmac.adb: $(srcdir)/ada/s-excmac-$(EH_MECHANISM).adb
 	$(CP) $< $@
 
 # Needs to be built with CC=gcc
diff --git a/gcc/ada/raise-gcc.c b/gcc/ada/raise-gcc.c
index cb35842b061..1264a726c63 100644
--- a/gcc/ada/raise-gcc.c
+++ b/gcc/ada/raise-gcc.c
@@ -32,12 +32,20 @@
 /* Code related to the integration of the GCC mechanism for exception
    handling.  */
 
-#ifndef CERT
-#include "tconfig.h"
-#include "tsystem.h"
+#ifndef IN_RTS
+  /* For gnat1/gnatbind compilation: use host headers.  */
+# include "config.h"
+# include "system.h"
+  /* Don't use fancy_abort.  */
+# undef abort
 #else
-#define ATTRIBUTE_UNUSED __attribute__((unused))
-#define HAVE_GETIPINFO 1
+# ifndef CERT
+#  include "tconfig.h"
+#  include "tsystem.h"
+# else
+#  define ATTRIBUTE_UNUSED __attribute__((unused))
+#  define HAVE_GETIPINFO 1
+# endif
 #endif
 
 #include <stdarg.h>
@@ -71,7 +79,19 @@ typedef char bool;
    (SJLJ or DWARF). We need a consistently named interface to import from
    a-except, so wrappers are defined here.  */
 
-#include "unwind.h"
+#ifndef IN_RTS
+  /* For gnat1/gnatbind compilation: cannot use unwind.h, as it is for the
+     target. So mimic configure...
+     This is a hack ???, the real fix is to link gnat1/gnatbind with the
+     runtime of the build compiler.  */
+# ifdef EH_MECHANISM_arm
+#   include "config/arm/unwind-arm.h"
+# else
+#   include "unwind-generic.h"
+# endif
+#else
+# include "unwind.h"
+#endif
 
 #ifdef __cplusplus
 extern "C" {
@@ -98,6 +118,11 @@ extern void __gnat_raise_abort (void) __attribute__ ((noreturn));
 
 #include "unwind-pe.h"
 
+#ifdef __ARM_EABI_UNWINDER__
+/* for memcmp */
+#include <string.h>
+#endif
+
 /* The known and handled exception classes.  */
 
 #ifdef __ARM_EABI_UNWINDER__
-- 
2.21.0.windows.1

